qt/python/mantidqt/project/encoderfactory.py

```diff
@@ -19,15 +19,23 @@ def find_encoder(cls, obj):
         :param obj: The object for encoding
         :return: Encoder or None; Returns the Encoder of the obj or None.
         """
-        for encoder in cls.encoder_list:
+        for encoder, compatible_check in cls.encoder_list:
+
+            # Perform check if optional_cb is present if it is, then check whether the the optional_cb returns true for
+            # this object.
+            if compatible_check is not None and compatible_check(obj):
+                return encoder()
+
             if encoder().has_tag(obj.__class__.__name__):
                 return encoder()
         return None
 
     @classmethod
-    def register_encoder(cls, encoder):
```

Comment:
How about something like this. Have a separate function:
```python
def default_encoder_compatability_check(encoder_cls, obj):
   return encoder_cls.has_tag(obj.__class__.__name__)

and then in `register_encoder`:
```python
if compatible_check is None:
    compatible_check = default_encoder_compatability_check
cls.encoder_list.add((encoder, compatible_check))
```
then your `find_encoder` doesn't ever have to deal with a `None` compatibility function
```python
obj_encoders = filter(lambda encoder, compatible: compatible(encoder), cls.encoder_list)
if len(obj_encoders) == 1:
   return obj_encoders[0]
else:
   raise RuntimeError("Some error message")
```
We can then also detect if two encoders claim to encode the same type which should be a bug?


 and its clearer here what the default check behaviour is.

Reply:
Having 0 encoders should also be ok. It is only an error if there is more than one, as many QWidget Object are pushed through this function.

Reply:
@Pasarus mentioned that `filter` doesn't seem to unpack the tuple, so this should do it `[encoder for encoder, compatible in cls.encoder_list if compatible(encoder)]`